include/swift/AST/DiagnosticEngine.h

```diff
@@ -629,7 +629,27 @@ namespace swift {
     DiagnosticState(DiagnosticState &&) = default;
     DiagnosticState &operator=(DiagnosticState &&) = default;
   };
-    
+
+  class LocalizationProducer {
+  public:
+    /// If the  message isn't available/localized in the current `yaml` file,
+    /// return the fallback default message.
+    virtual std::string getMessageOr(DiagID id,
+                                     std::string defaultMessage) const {
+      return defaultMessage;
+    }
+
+    virtual ~LocalizationProducer() {}
+  };
+
+  class YAMLLocalizationProducer final : public LocalizationProducer {
+  public:
+    std::vector<std::string> diagnostics;
```

Comment:
`llvm::SmallVector` maybe?
@xedin seems the pattern across the project to use `llvm` structures, but that is something I'm curious about is where it would be ok to use STL types e.g. `std::vector/set ...`? :)

Reply:
This is really a good question, waiting for the answer 😅

Reply:
This came up recently when I was refactoring some SIL vtable code. It seems as if one cannot reasonably guess what size a SmallVector should be, then a std::vector should be fine. (In my case, it would have been weird to use a `llvm::SmallVector` to represent the number of vtables per module.)

Reply:
Ah, thanks @davezarzycki :) So for this case an `std::vector` for the diagnostics should be fine too.

Reply:
@LucianoPAlmeida There are going to be multiple thousands of diagnostics in this list, no real reason to spent stack space which would have to be re-allocated anyway once list grows over hundred items or so... `SmallVector` is very useful in situations when there are usually not too many items in the list to avoid heap allocations.

Reply:
Ah, TIL that `SmallVector` is stack-allocated, that makes total sense. Thank you @xedin :)

Reply:
@LucianoPAlmeida Yeah, that number in generic specialization is used to figure out how much stack to allocate, if number of items doesn't exceed specified maximum `SmallVector` stays on stack.

Reply:
Humm, that is why if one can guess upfront the size or possible max size(if small) which are the majority of cases it is indeed the way to go. That's great, thanks for the explanation @xedin :)